### PR TITLE
Cleanup and refactor public API in gosnappi SDK

### PR DESCRIPTION
**Pull request by Vibaswan**
### How to consume this branch ?

```
go get github.com/open-traffic-generator/snappi/gosnappi@refactor-go-sdk
```

### Who all are impacted by these changes ?

- ONDATRA
- FeatureProfiles
- OTG implementations using gosnappi
  - https://github.com/open-traffic-generator/ixia-c-gnmi-server
  - keng-controller
  - lemming / magna ?
- Community tests hosted under OTG organization

### What are the changes ?

 * `GosnappiApi` interface is now renamed to `Api` interface.
   * Maybe used in function signatures
   * Breaks backward compatibility
       
     Before:
     ```go
     type GosnappiApi interface {
        api
        // SetConfig sets configuration resources on the traffic generator.
        SetConfig(config Config) (Warning, error)
       // GetConfig description is TBD
        GetConfig() (Config, error)
        ...
      ```
  
     After:
     ```go
     type Api interface {
        api
        // SetConfig sets configuration resources on the traffic generator.
        SetConfig(config Config) (Warning, error)
        // GetConfig description is TBD
        GetConfig() (Config, error)
        ...
     ```
*  All public methods for creation of structs are now removed from `GosnappiApi` interface
   * Those public methods are only accessible on package level
   * Breaks backward compatibility
      
      Before:
      ```go
        // the following way of creating structs were possible
        api := gosnappi.NewApi()
        config := api.NewConfig()
        ...
     ```

     After:
     ```go
        config := gosnappi.NewConfig()
        ...
      ```    
    * `FromError()` method is now exposed on package level instead of `api` level
      * Breaks backward compatibility
      
      Before:
      ```go
       api := gosnappi.NewApi()
       config := api.NewConfig()
       ....
       err := ...
       _, err = api.FromError(err)
      ```

      After:     
      ```go
       config := gosnappi.NewConfig()
       ....
       err := ...
       _, err = gosnappi.FromError(err)
      ```

    * `Warnings()` method is now private 
      * Might be exposed later again, in a slightly different form, if use cases arise
      * Breaks backward compatibility
      
      Before:     
      ```go
      // earlier warnings was exposed at api level
      api := gosnappi.NewApi()
      ...
      fmt.Println(api.Warnings()) 
      ```
     
      After: 
      ```go
      // NA
      ```

* There are helper methods defined on each struct which have been reorganized or hidden
    * All marshalling/unmarshalling helpers are now moved behind two methods `Marhsal()` and `Unmarshal()`
      * Breaks backward compatibility
       
      Before:
       ```go
       api := gosnappi.NewApi()
       config := api.NewConfig()
       ...
       str, err := config.ToJson()
       str, err = config.ToYaml()
       str, err =config.ToPbText()
       pr, err = config.ToProto()
       err = config.FromJson(str)
       err = config.FromYaml(str)
       err = config.FromPbText(str)
       conf, err = config.FromProto(pr)
      ```
     
      After:
      ```go
       config := gosnappi.NewConfig()
       ...
       str, err := config.Marshal().ToJson()
       str, err = config.Marshal().ToYaml()
       str, err = config.Marshal().ToPbText()
       pr, err = config.Marshal().ToProto()
       err = config.Unmarshal().FromJson(str)
       err = config.Unmarshal().FromYaml(str)
       err = config.Unmarshal().FromPbText(str)
       conf, err = config.Unmarhsal().FromProto(pr)
      ```
      
    * `Msg()`, `SetMsg()` and `Validate()` are now private
      * Breaks backward compatibility
       
       Before:      
       ```go
        api := gosnappi.NewApi()
        config := api.NewConfig()
        ...
        msg = config.Msg()
        config.setMsg(msg) 
       ```

       After:
       ```go
          config := gosnappi.NewConfig()
          ...
          msg, err = config.Marshal().ToProto()
          conf, err = config.Unmarshal().FromProto(msg) 
       ```
    
   * `Clone()` and `String()` will remain as is
    
        ```go
         config := gosnappi.NewConfig()
         ...
         fmt.Println(config.String())
         conf, err := config.Clone()
         ...
        ```

*  Choice setter `SetChoice()` was leading to _“two ways of doing the same thing”_ and hence has been made private
  * In gosnappi, choice is supposed to be implicitly set based on the choice property set by the user
  * Breaks backward compatibility    
   
    
    Before:
    ```go
     api := gosnappi.NewApi()
     config := api.NewConfig()
     port1 := config.Ports().Add()
     f1 := config.Flows().Add().SetName("f1")
     txRx := f1.TxRx()
     // explicit call for set choice allowed although not needed
     txRx.SetChoice(gosnappi.FlowTxRxChoice.PORT)
     // when we call port it sets the choice to port if not already set, thats why setchoice need not be called explicitly
     txRx.Port().SetTxName(port1.Name())
    ```

     After:
    ```go
     config := gosnappi.NewConfig()
     port1 := config.Ports().Add()
     f1 := config.Flows().Add().SetName("f1")
     f1.TxRx().Port().SetTxName(port1.Name())
     // txRx.SetChoice not allowed
    ```

### CODE REVIEW

File: gosnappi/httpapi/controllers/capabilities_controller.go [hunk 1]
```diff
@@ -58,18 +58,18 @@ func (ctrl *capabilitiesController) responseGetVersionError(w http.ResponseWrite
 		result = rErr
 	} else {
 		result = gosnappi.NewError()
-		err := result.FromJson(rsp_err.Error())
+		err := result.Unmarshal().FromJson(rsp_err.Error())
 		if err != nil {
-			result.Msg().Code = &statusCode
+			_ = result.SetCode(statusCode)
```
Comment (greg-dennis) [inline gosnappi/httpapi/controllers/capabilities_controller.go:63]: Go doesn't requite "_ =" if you don't want the noise. Up to you.

**Comment by greg-dennis** [comment]
Can you fix your description above to replace "Unmarhsal" with "Unmarshal"?

**Review by greg-dennis — APPROVED** [review]
One possible nit. All of this looks good to me.

File: gosnappi/common.go [hunk 2]
```diff
@@ -116,31 +115,28 @@ func (obj *httpTransport) SetVerify(value bool) HttpTransport {
 	return obj
 }
 
-type api struct {
+type apiSt struct {
```
Comment (ashutshkumr) [inline gosnappi/common.go:118]: maybe `apiImpl`, since `apiSt` sounds like api status ?

Can be addressed after merge.

**Review by ashutshkumr — APPROVED** [review]

**Comment by ashutshkumr** [comment]
I would ideally rename `marshalConfig` -> `configMarshaller`, etc.

Can be addressed after the merge.

```go
type config struct {
	validation
	obj            *otg.Config
	marshaller     marshalConfig
	unMarshaller   unMarshalConfig
```

**Pull request merged**